test/extensions/filters/http/stateful_session/stateful_session_integration_test.cc

```diff
@@ -0,0 +1,393 @@
+
+  upstream_request_->encodeHeaders(default_response_headers_, true);
+
+  ASSERT_TRUE(response->waitForEndStream());
+
+  EXPECT_TRUE(upstream_request_->complete());
+  EXPECT_TRUE(response->complete());
+
+  // The selected upstream server address would be selected to the response headers.
+  EXPECT_EQ(
+      Envoy::Http::Utility::makeSetCookieValue("global-session-cookie", encoded_address, "/path",
+                                               std::chrono::seconds(120), true),
+      response->headers().get(Http::LowerCaseString("set-cookie"))[0]->value().getStringView());
+
+  cleanupUpstreamAndDownstream();
+}
+
+TEST_F(StatefulSessionIntegrationTest, DownstreamRequestWithStatefulSessionCookie) {
+  initializeFilterAndRoute(STATEFUL_SESSION_FILTER, "");
+
+  {
+    codec_client_ = makeHttpConnection(lookupPort("http"));
+    Http::TestRequestHeaderMapImpl request_headers{
+        {":method", "GET"},
+        {":path", "/test"},
+        {":scheme", "http"},
+        {":authority", "stateful.session.com"},
+        {"cookie", fmt::format("global-session-cookie=\"{}\"",
+                               Envoy::Base64::encode("127.0.0.1:50001", 15))}};
+
+    auto response = codec_client_->makeRequestWithBody(request_headers, 0);
+
+    // `127.0.0.1:50001` should be selected and it's upstream index is 1.
+    auto upstream_index = waitForNextUpstreamRequest({0, 1, 2, 3});
+    EXPECT_EQ(upstream_index.value(), 1);
+
+    upstream_request_->encodeHeaders(default_response_headers_, true);
+
+    ASSERT_TRUE(response->waitForEndStream());
+
+    EXPECT_TRUE(upstream_request_->complete());
+    EXPECT_TRUE(response->complete());
+
+    // No response header to be added.
+    EXPECT_TRUE(response->headers().get(Http::LowerCaseString("set-cookie")).empty());
+
+    cleanupUpstreamAndDownstream();
+  }
+
+  {
+    codec_client_ = makeHttpConnection(lookupPort("http"));
+    Http::TestRequestHeaderMapImpl request_headers{
+        {":method", "GET"},
+        {":path", "/test"},
+        {":scheme", "http"},
+        {":authority", "stateful.session.com"},
+        {"cookie", fmt::format("global-session-cookie=\"{}\"",
+                               Envoy::Base64::encode("127.0.0.1:50002", 15))}};
+
+    auto response = codec_client_->makeRequestWithBody(request_headers, 0);
+
+    // `127.0.0.1:50002` should be selected and it's upstream index is 2.
+    auto upstream_index = waitForNextUpstreamRequest({0, 1, 2, 3});
+    EXPECT_EQ(upstream_index.value(), 2);
+
+    upstream_request_->encodeHeaders(default_response_headers_, true);
+
+    ASSERT_TRUE(response->waitForEndStream());
+
+    EXPECT_TRUE(upstream_request_->complete());
+    EXPECT_TRUE(response->complete());
+
+    // No response header to be added.
+    EXPECT_TRUE(response->headers().get(Http::LowerCaseString("set-cookie")).empty());
+
+    cleanupUpstreamAndDownstream();
+  }
+
+  // Test the case that stateful session cookie with unknown server address.
```

Comment:
@dio  And there is a test case with unknown server address. Envoy will just select new one from the exist host list and ignore this unknown address.

Reply:
Got it thanks for pointing it out!